mobile/examples/phi-3/ios/LocalLLM/LocalLLM/GenAIGenerator.mm

```diff
@@ -1,143 +1,146 @@
-        NSString* errorMessage = e.reason;
-        NSLog(@"Error during generation: %@", errorMessage);
-
-        // Send error to the UI
-        NSDictionary *errorInfo = @{@"error": errorMessage};
-        dispatch_async(dispatch_get_main_queue(), ^{
-            [[NSNotificationCenter defaultCenter] postNotificationName:@"TokenGenerationError" object:nil userInfo:errorInfo];
-        });
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    self->model = nullptr;
+    self->tokenizer = nullptr;
+  }
+  return self;
+}
+
+- (void)generate:(nonnull NSString*)input_user_question {
+  std::vector<long long> tokenTimes;  // per-token generation times
+  TimePoint startTime, firstTokenTime, tokenStartTime;
+
+  try {
+    NSLog(@"Starting token generation...");
+
+    if (!self->model) {
+      NSLog(@"Creating model...");
+      NSString* llmPath = [[NSBundle mainBundle] resourcePath];
+      const char* modelPath = llmPath.cString;
+      self->model = OgaModel::Create(modelPath);  // throws exception
+    }
+
+    if (!self->tokenizer) {
+      NSLog(@"Creating tokenizer...");
+      self->tokenizer = OgaTokenizer::Create(*self->model);  // throws exception
+    }
+
+    auto tokenizer_stream = OgaTokenizerStream::Create(*self->tokenizer);
+
+    // Construct the prompt
+    NSString* promptString = [NSString stringWithFormat:@"<|user|>\n%@<|end|>\n<|assistant|>", input_user_question];
+    const char* prompt = [promptString UTF8String];
+
+    // Encode the prompt
+    auto sequences = OgaSequences::Create();
+    self->tokenizer->Encode(prompt, *sequences);
+
+    size_t promptTokensCount = sequences->SequenceCount(0);
+
+    NSLog(@"Setting generator parameters...");
+    auto params = OgaGeneratorParams::Create(*self->model);
+    params->SetSearchOption("max_length", 200);
+    params->SetInputSequences(*sequences);
+
+    auto generator = OgaGenerator::Create(*self->model, *params);
+
+    bool isFirstToken = true;
+    NSLog(@"Starting token generation loop...");
+
+    startTime = Clock::now();
+    while (!generator->IsDone()) {
+      tokenStartTime = Clock::now();
+
+      generator->ComputeLogits();
+      generator->GenerateNextToken();
+
+      if (isFirstToken) {
+        firstTokenTime = Clock::now();
+        isFirstToken = false;
+      }
+
+      // Get the sequence data and decode the token
+      const int32_t* seq = generator->GetSequenceData(0);
+      size_t seq_len = generator->GetSequenceCount(0);
+      const char* decode_tokens = tokenizer_stream->Decode(seq[seq_len - 1]);
+
+      if (!decode_tokens) {
+        @throw [NSException exceptionWithName:@"TokenDecodeError" reason:@"Token decoding failed." userInfo:nil];
+      }
+
+      // Measure token generation time excluding logging
+      TimePoint tokenEndTime = Clock::now();
+      auto tokenDuration = std::chrono::duration_cast<std::chrono::milliseconds>(tokenEndTime - tokenStartTime).count();
+      tokenTimes.push_back(tokenDuration);
```

Comment:
nit: consider calling `tokenTimes.reserve()` first to avoid `std::vector` reallocations during the measurement.